File: app/api/v1/entities/workspace.rb
# frozen_string_literal: true

module V1
  module Entities
    class Workspace < Grape::Entity
      expose :id
      expose :name
      expose :slug
      expose :workspace_type
      expose :plan
      expose :logo_url
      expose :description
      expose :is_active
      expose :created_at
      expose :updated_at

      # Owner info
      expose :owner_id
      expose :owner, using: V1::Entities::User, if: ->(workspace, options) { options[:include_owner] }

      # Current user's role in this workspace
      expose :current_user_role do |workspace, options|
        next unless options[:current_user]

        workspace.role_for(options[:current_user])
      end

      # Member count
      expose :member_count do |workspace|
        workspace.workspace_memberships.active.count
      end

      # Is this the user's personal workspace?
      expose :is_personal do |workspace|
        workspace.personal?
      end

      # Is the current user the owner?
      expose :is_owner do |workspace, options|
        next false unless options[:current_user]

        workspace.owner?(options[:current_user])
      end
    end

    class WorkspaceSimple < Grape::Entity
      expose :id
      expose :name
      expose :slug
      expose :workspace_type
      expose :logo_url
      expose :is_active

      expose :current_user_role do |workspace, options|
        next unless options[:current_user]

        workspace.role_for(options[:current_user])
      end
    end
  end
end

File: app/javascript/types/workspace.ts
export type WorkspaceType = 'personal' | 'business';
export type WorkspaceRole = 'owner' | 'admin' | 'accountant' | 'member' | 'viewer';
export type MembershipStatus = 'pending' | 'active' | 'suspended';

export interface Workspace {
  id: number;
  name: string;
  slug: string;
  workspace_type: WorkspaceType;
  plan: string;
  logo_url: string | null;
  description: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  owner_id: number;
  current_user_role: WorkspaceRole | null;
  member_count: number;
  is_personal: boolean;
  is_owner: boolean;
}

export interface WorkspaceSimple {
  id: number;
  name: string;
  slug: string;
  workspace_type: WorkspaceType;
  logo_url: string | null;
  is_active: boolean;
  current_user_role: WorkspaceRole | null;
}

export interface WorkspaceMember {
  id: number;
  user_id: number;
  role: WorkspaceRole;
  status: MembershipStatus;
  joined_at: string | null;
  last_accessed_at: string | null;
  created_at: string;
  user: {
    id: number;
    name: string | null;
    email: string | null;
    avatar_url: string | null;
  };
  is_owner: boolean;
  is_admin: boolean;
}

export interface WorkspaceInvitation {
  id: number;
  email: string | null;
  phone_number: string | null;
  role: WorkspaceRole;
  status: string;
  expires_at: string;
  created_at: string;
  invited_by: {
    id: number;
    name: string | null;
  };
  is_expired: boolean;
}

export interface CreateWorkspaceData {
  name: string;
  workspace_type: WorkspaceType;
  description?: string;
  logo_url?: string;
}

export interface InviteMemberData {
  email?: string;
  phone_number?: string;
  role: Exclude<WorkspaceRole, 'owner'>;
}

export interface UpdateMemberData {
  role: Exclude<WorkspaceRole, 'owner'>;
}
